urllib3/sessionmanager.py

```diff
@@ -0,0 +1,78 @@
+from .util.retry import Retry, handle_redirect
+from .request import RequestMethods, Request
+from .poolmanager import PoolManager, ProxyManager
+from .util.context import Context
+
+
+class SessionManager(RequestMethods):
+    """
+    Allows arbitrary requests while maintaining browser-session-like
+    context across those requests.
+
+    :param headers:
+        Headers to include with all requests, unless other headers are
+        given explcitly
+
+    :param cookie_jar:
+        A predefined :class:`CookieJar` object to use in the session;
+        if not provided, a new one will be created.
+
+    :param \**pm_kwargs:
+        Additional paramaters to be passed to the encapsulated
+        :class:`.PoolManager` or :class:`.ProxyManager` instance.
+    """
+    def __init__(self, headers=None, cookie_jar=None, **pm_kwargs):
+
+        super(SessionManager, self).__init__(headers=headers)
+
+        # Detect if we should create a PoolManager or a ProxyManager
+        if pm_kwargs.get('proxy_url', None) is None:
```

Comment:
Primarily because the coupling in PoolManager was unavoidable since it needs to create new instances of ConnectionPool per host. Whereas SessionManager is 1:1 so we can avoid it and it should make the code cleaner by avoiding switches based on input conditions. For example, there may come a point when PoolManager and ProxyManager or OtherFutureManager have the same kw param with conflicting meanings, and then it will be impossible to create a SessionManager with the appropriate type of internal manager.

I'm all for shortcuts (we have `connection_from_*` and similar all over the place for this), but I feel like constructors should be as decoupled as possible to allow better substitution.

Also, there's no reason a user shouldn't be allowed to provide their own `CustomPoolManager` as the driver for the `SessionManager`.